src/covid_model_seiir_pipeline/marshall.py

```diff
@@ -0,0 +1,442 @@
+
+    @classmethod
+    def date(cls, draw_id):
+        return cls(DataTypes.date, DRAW_FILE_TEMPLATE, draw_id=draw_id)
+
+    @classmethod
+    def fit_beta(cls, draw_id):
+        return cls(DataTypes.fit_beta, DRAW_FILE_TEMPLATE, draw_id=draw_id)
+
+    @classmethod
+    def parameter(cls, draw_id):
+        return cls(DataTypes.parameter, DRAW_FILE_TEMPLATE, draw_id=draw_id)
+
+    @classmethod
+    def regression_beta(cls, draw_id):
+        return cls(DataTypes.regression_beta, DRAW_FILE_TEMPLATE, draw_id=draw_id)
+
+    # factories related to forecasting
+    @classmethod
+    def beta_scales(cls, scenario, draw_id):
+        return cls(DataTypes.beta_scales, DRAW_FILE_TEMPLATE, scenario=scenario, draw_id=draw_id)
+
+    @classmethod
+    def components(cls, scenario, draw_id):
+        return cls(DataTypes.components, DRAW_FILE_TEMPLATE, scenario=scenario, draw_id=draw_id)
+
+    # other methods/properties
+    @property
+    def key(self):
+        return self.template.format(**self.key_args)
+
+    @property
+    def seed(self):
+        """
+        Returns a seed value for enabling concurrency.
+
+        This is a sort of hack - we embed the concurrency story into the Keys
+        class which can then inform other things.
+        """
+        assert list(self.key_args) == ['draw_id'], 'TODO - expand Keys.seed'
+        return "draw-{draw_id}".format(**self.key_args)
+
+    def __repr__(self):
+        return f"Keys({self.data_type!r}, {self.template!r}, **{self.key_args!r})"
+
+
+class CSVMarshall:
+    """
+    Marshalls DataFrames to/from CSV files.
+
+    This implementation directly mirrors existing behavior but does so within a
+    new marshalling interface.
+    """
+    # interface methods
+    def dump(self, data: pandas.DataFrame, key):
+        path = self.resolve_key(key)
+        if not path.parent.is_dir():
+            mkdir(path.parent, parents=True)
+        else:
+            if path.exists():
+                msg = f"Cannot dump data for key {key} - would overwrite"
+                raise LookupError(msg)
+
+        data.to_csv(path, index=False)
+
+    def load(self, key):
+        path = self.resolve_key(key)
+        return pandas.read_csv(path)
+
+    def resolve_key(self, key):
+        if key.data_type in DataTypes.DataFrame_types:
+            scenario = key.key_args.get('scenario')
+            if scenario is None:
+                path = (self.root / key.data_type / key.key).with_suffix(".csv")
+            else:
+                path = (self.root / scenario / key.data_type / key.key).with_suffix(".csv")
+        else:
+            msg = f"Invalid 'type' of data: {key.data_type}"
+            raise ValueError(msg)
+
+        return path
+
+    @classmethod
+    def from_paths(cls, paths: Paths):
+        return cls(paths.root_dir)
+
+    # non-interface methods
+    def __init__(self, root: Path):
+        self.root = root
+
```

Comment:
The below classes (ZipMarshall and Hdf5Marshall) are *no longer working* because they do not yet handle nesting of forecasting data by scenario name (which I learned about/added in the last couple hours).

I am working on it now - it just requires updating the `resolve_key` methods to deal with that detail.

I'm not entirely happy with how the separation of file path logic worked out.

In this case the `Keys` class cannot be wholly responsible because it is not supposed to know about implementation details of CSV/Zip/HDF. However the Marshall class also cannot be wholly responsible without duplicating logic or creating a class hierarchy to share small amounts of information.

On the positive side all the ugliness of this is at least contained to a few functions all in one module...